### PR TITLE
[HttpKernel][VarDumper] Truncate profiler data & optim perf

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #23415, #21547 and hopefully #23110 and #23175
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/DataCollector/DataCollector.php [hunk 1]
```diff
@@ -100,4 +106,24 @@ protected function varToString($var)
...
         return $this->valueExporter->exportValue($var);
     }
+
+    /**
+     * @return callable[] The casters to add to the cloner
+     */
+    protected function getCasters()
+    {
+        return array(
+            '*' => function ($v, array $a, Stub $s, $isNested) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DataCollector/DataCollector.php]: should be the main fix: nested objects are truncated, which prevent deep recursion into nested structures by default (eg "subjects" of the security decision log, requests attributes, etc.)

File: src/Symfony/Component/VarDumper/Caster/Caster.php [hunk 2]
```diff
@@ -65,11 +65,20 @@ public static function castObject($obj, $class, $hasDebugInfo = false)
         }
 
         if ($a) {
+            static $publicProperties = array();
+
             $i = 0;
             $prefixedKeys = array();
             foreach ($a as $k => $v) {
-                if (isset($k[0]) && "\0" !== $k[0] && !property_exists($class, $k)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Caster/Caster.php]: as shown by profiles in linked issue, this is a hot code path and `property_exists()` calling should be prevented

File: src/Symfony/Component/VarDumper/Cloner/AbstractCloner.php [hunk 3]
```diff
@@ -210,14 +210,12 @@ public function cloneVar($var, $filter = 0)
         $this->filter = $filter;
 
         try {
+            gc_disable();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Cloner/AbstractCloner.php]: shown also in linked profiles: cloning involves lots of `Stub` objects, thus can trigger lots of GC calls, for no benefit.

**Comment by szymach** [comment]
Hey, I have just tried this fix and it indeed improves performance a lot. Basically it runs as fast as 3.2 did. Good job :)

File: src/Symfony/Component/VarDumper/Cloner/AbstractCloner.php (hunk 3 above)
Comment (ro0NL) [inline src/Symfony/Component/VarDumper/Cloner/AbstractCloner.php]: Shouldnt we only re-enable based on `gc_enabled()`?

File: src/Symfony/Component/Form/Extension/DataCollector/FormDataCollector.php [hunk 4]
```diff
@@ -255,61 +246,33 @@ public function serialize()
...
-        return $this->valueExporter->exportValue($var);
+                return $a;
+            },
+            FormInterface::class => function (FormInterface $f, array $a) {
+                return array(
+                    Caster::PREFIX_VIRTUAL.'name' => $f->getName(),
+                    Caster::PREFIX_VIRTUAL.'type_class' => new ClassStub(get_class($f->getConfig()->getType()->getInnerType())),
+                );
+            },
+            ConstraintViolationInterface::class => function (ConstraintViolationInterface $v, array $a) {
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/DataCollector/FormDataCollector.php]: should we add this caster in the new Validator panel in 3.4 too ?

File: src/Symfony/Component/VarDumper/Cloner/AbstractCloner.php (hunk 3 above)
Comment (stof) [inline src/Symfony/Component/VarDumper/Cloner/AbstractCloner.php]: agreed

Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Cloner/AbstractCloner.php]: fixed

File: src/Symfony/Component/Form/Extension/DataCollector/FormDataCollector.php (hunk 4 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/DataCollector/FormDataCollector.php]: that's for you @ogizanagi

File: src/Symfony/Component/HttpKernel/DataCollector/DataCollector.php [hunk 5]
```diff
@@ -100,4 +105,24 @@ protected function varToString($var)
...
+                if (!$v instanceof Stub) {
+                    foreach ($a as &$v) {
```
Comment (stof) [inline src/Symfony/Component/HttpKernel/DataCollector/DataCollector.php]: @nicolas-grekas have you checked whether the usage of a reference here could be the reason why https://github.com/symfony/symfony/issues/23110 happens (replacing the EntityManager somewhere it should not be replaced) ?

File: src/Symfony/Component/Form/Extension/DataCollector/FormDataCollector.php (hunk 4 above)
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/DataCollector/FormDataCollector.php]: Sure, I'll check this once this is merge on 3.4.

File: src/Symfony/Component/HttpKernel/DataCollector/DataCollector.php (hunk 5 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DataCollector/DataCollector.php]: I removed the reference so that *if* this is correct, then it might be fixed

**Review by stof — APPROVED** [review]

**Review by ogizanagi — APPROVED** [review]

**Review by szymach — APPROVED** [review]

**Comment by sroze** [comment]
Could we have an opt-in for this? As we can see in a few issues, this loses some information that might be useful to the developer. It could be nice to be able to "tune" a "deep debug" or "simple debug" something like that... 🤔

**Pull request merged**